doc/man3/OSSL_SERIALIZER_to_bio.pod

```diff
@@ -0,0 +1,94 @@
+=pod
+
+=head1 NAME
+
+OSSL_SERIALIZER_to_bio,
+OSSL_SERIALIZER_to_fp,
+OSSL_SERIALIZER_CTX_set0_object,
+OSSL_SERIALIZER_CTX_get0_object,
+OSSL_SERIALIZER_CTX_set_do_output
+- Serializer file output routines
+
+=head1 SYNOPSIS
+
+ #include <openssl/serializer.h>
+
+ int OSSL_SERIALIZER_to_bio(OSSL_SERIALIZER_CTX *ctx, BIO *out);
+ int OSSL_SERIALIZER_to_fp(OSSL_SERIALIZER_CTX *ctx, FILE *fp);
+
+ int OSSL_SERIALIZER_CTX_set0_object(OSSL_SERIALIZER_CTX *ctx,
+                                     const void *object);
+ const void *OSSL_SERIALIZER_CTX_get0_object(OSSL_SERIALIZER_CTX *ctx);
+ int OSSL_SERIALIZER_CTX_set_do_output(OSSL_SERIALIZER_CTX *ctx,
+                                       int (*do_output)(OSSL_SERIALIZER_CTX *ctx,
+                                                        BIO *out));
+
+Feature availability macros:
+
+=over 4
+
+=item OSSL_SERIALIZER_to_fp() is only available with B<OPENSSL_NO_STDIO>
+is undefined.
+
+=back
+
+=head1 DESCRIPTION
+
+OSSL_SERIALIZER_to_bio() runs the serialization process for the
+context I<ctx>, with the output going to the B<BIO> I<out>.  The
+application is required to set up the B<BIO> properly, for example to
+have it in text or binary mode if that's appropriate.
+
+=for comment Know your serializer!
+
+OSSL_SERIALIZER_to_fp() does the same thing as OSSL_SERIALIZER_to_bio(),
+except that the output is going to the B<FILE> I<fp>.
+
+OSSL_SERIALIZER_CTX_set0_object() sets the object to be serialized in
+the context.  This object must be intimately known by the processing
+function gitven with OSSL_SERIALIZER_CTX_set_do_output() (see it's
+description below).  No limit is imposed on the passed I<object>, so
+NULL is a valid value.
+
+OSSL_SERIALIZER_CTX_get0_object() gets back the object that was set in
+I<ctx> with OSSL_SERIALIZER_CTX_set0_object().
+
+OSSL_SERIALIZER_CTX_set_do_output() adds the pointer to a serialization
```

Comment:
I hadn't thought of that, but if some application wants to make a serializer of their own for some type, sure why not?

Reply:
Well...if that was not your original intention...how were you expecting this to be used? Does it belong on the public API?

If some application does that it reminds me of the legacy _meth_ APIs that we have that we want to move away from, so I'm not sure we should be adding new APIs to do something very similar.

Reply:
should this use the exdata facility to be consistent and generic?

Reply:
If you guys look at the implementation of `OSSL_SERIALIZER_CTX_new_by_EVP_PKEY`, you'll see how I *intend* these to be used.

My thought was that anyone could add support to serialize some other type, something like this for example:

``` C
static int serialize_x509_to_bio(OSSL_SERIALIZER_CTX *ctx, BIO *out)
{
    const X509 *x509 = OSSL_SERIALIZER_CTX_get0_object(ctx);

    if (ctx->ser == NULL)
        return 0;

    /*
     * ... do whatever's needed to call ctx->ser->serialize_object or
     * ctx->ser->serialize_data with required arguments
     */
}

OSSL_SERIALIZER_CTX *OSSL_SERIALIZER_CTX_new_by_X509(const X509 *x509,
                                                     const char *propquery)
{
    OSSL_SERIALIZER *ser = OSSL_SERIALIZER_fetch(NULL, "X509", propquery);

    ctx = OSSL_SERIALIZER_CTX_new(ser); /* refcnt(ser)++ */
    OSSL_SERIALIZER_free(ser);          /* refcnt(ser)-- */

    if (ctx != NULL) {
        OSSL_SERIALIZER_CTX_set0_object(ctx, x509);
        OSSL_SERIALIZER_CTX_set_do_output(ctx, serializer_x509_to_bio);
    }

    return ctx;
}
```

I'm thinking that if *someone else* wants to write support for serializing types that we haven't written support for, why not?

`OSSL_SERIALIZER_CTX_set_do_output()` is *nothing* like method building.  It sets a function pointer in a context, i.e. is a fly-by thing rather than a method building thing.  It's close to the diverse function setter you find in `include/openssl/x509_vfy.h`, like `X509_STORE_set_verify`

Reply:
I've had a think on these functions, and have reconsidered.  They will become internal.

Reply:
... or actually, disappear.  I was going to make them local only, and consider the context structure is available locally anyway, it became ridiculous to go via functions.